rasa/engine/caching.py

```diff
@@ -0,0 +1,373 @@
+                fingerprints.
+
+        Returns:
+            The fingerprint of a matching output or `None` in case no cache entry was
+            found for the given fingerprint key.
+        """
+        with self._sessionmaker.begin() as session:
+            query = sa.select(self.CacheEntry).filter_by(
+                fingerprint_key=fingerprint_key
+            )
+            match = session.execute(query).scalars().first()
+
+            if match:
+                # This result was used during a fingerprint run.
+                match.last_used = datetime.utcnow()
+                return match.output_fingerprint_key
+
+            return None
+
+    def get_cached_result(
+        self, output_fingerprint_key: Text
+    ) -> Tuple[Optional[Path], Optional[Type[Cacheable]]]:
+        """Returns a potentially cached output result.
+
+        Args:
+            output_fingerprint_key:
+
+        Returns:
+            - Path to directory containing the cached content or `None` in case no
+            persisted data was found
+            - Class of persisted data or None in case no persisted data was found.
+        """
+        result_location, result_type = self._get_cached_result(output_fingerprint_key)
+
+        if not result_location:
+            logger.debug(f"No cached output found for '{output_fingerprint_key}'")
+            return None, None
+
+        path_to_cached = Path(result_location)
+        if not path_to_cached.is_dir():
+            logger.debug(
+                f"Cached output for '{output_fingerprint_key}' can't be found on disk."
+            )
+            return None, None
+
+        return self._load_from_cache(result_location, result_type)
+
+    def _get_cached_result(
+        self, output_fingerprint_key: Text
+    ) -> Tuple[Optional[Path], Optional[Text]]:
+        with self._sessionmaker.begin() as session:
+            query = sa.select(
+                self.CacheEntry.result_location, self.CacheEntry.result_type
+            ).where(
+                self.CacheEntry.output_fingerprint_key == output_fingerprint_key,
+                self.CacheEntry.result_location != sa.null(),
+            )
+
+            match = session.execute(query).first()
+
+            if match:
+                return Path(match.result_location), match.result_type
+
+            return None, None
+
+    @staticmethod
+    def _load_from_cache(
+        path_to_cached: Path, result_type: Text
+    ) -> Tuple[Optional[Path], Optional[Type[Cacheable]]]:
+        try:
+            module = rasa.shared.utils.common.class_from_module_path(result_type)
+            assert isinstance(module, Cacheable)
+
+            return path_to_cached, module
+        except Exception as e:
+            logger.warning(
+                f"Failed to find module for cached output of type "
+                f"'{result_type}'. Error:\n{e}"
+            )
+            return None, None
+
+
+def _directory_size_in_mb(path: Path) -> float:
```

Comment:
Is there no existing util to do this?

Reply:
unfortunately not 😬

Reply:
might be worth making it a public util method and unit testing it.

Reply:
will do 👍🏻

Reply:
done